src/osp/OneStepProverHostIo.sol

```diff
@@ -158,6 +177,47 @@ contract OneStepProverHostIo is IOneStepProver {
                 preimageEnd = preimage.length;
             }
             extracted = preimage[preimageOffset:preimageEnd];
+        } else if (inst.argumentData == 2) {
+            // The machine is asking for an Ethereum versioned hash preimage
+
+            require(proofType == 0, "UNKNOWN_PREIMAGE_PROOF");
+
+            bytes calldata kzgProof = proof[proofOffset:];
+
+            require(bytes32(kzgProof[:32]) == leafContents, "KZG_PROOF_WRONG_HASH");
+
+            (bool success, bytes memory kzgParams) = address(0x0A).staticcall(kzgProof);
+            require(success, "INVALID_KZG_PROOF");
+            require(kzgParams.length > 0, "KZG_PRECOMPILE_MISSING");
+            (uint256 fieldElementsPerBlob, ) = abi.decode(kzgParams, (uint256, uint256));
+
+            // If preimageOffset is greater than the blob size, leave extracted empty and call it here.
+            if (preimageOffset < fieldElementsPerBlob * 32) {
```

Comment:
Why do we return empty here instead of reverting?

Reply:
This is according to the API of the ReadPreImage opcode. If the offset requested is past the end of the preimage, an empty byte array is returned. That's so that the WASM can figure out the length of the preimage. The replay binary keeps calling ReadPreImage at 32 byte increments until it reaches a preimage of size <32. Theoretically, we could put the machine in the error state if preimageOffset > fieldElementsPerBlob * 32, but the API seems simpler to me if we just return an empty byte array in that case.